npc-utils.js

```diff
@@ -0,0 +1,528 @@
+  }
+
+  async init() { // Use highestY and highestY2, to auto rise() -> rise2() -> generateVoxelMap().
+    const re = await new Promise(resolve => {
+      this.resolveInit = resolve;
+
+      this.isAutoInit = true;
+      this.rise();
+      this.update();
+    });
+    if (re === true) console.log('PathFinder auto inited.');
+  }
+
+  update() {
+    // console.log('update');
+    if (this.isRising && this.voxels) { // mark: generate voxel map
+      this.voxels.children.forEach((voxel, i) => {
+        if (voxel._risingState === 'initial' || voxel._risingState === 'colliding') {
+          voxel.position.y += 0.1;
+          voxel.updateMatrixWorld();
+          // const isCollide = physicsManager.collideCapsule(0.5, 1, voxel.position, localQuaternion.set(0, 0, 0, 1), 1);
+          // const isCollide = physicsManager.collideBox(0.5, 0.05, 0.5, voxel.position, localQuaternion.set(0, 0, 0, 1), 1);
+          const isCollide = physicsManager.overlapBox(0.5, this.voxelHeightHalf, 0.5, voxel.position, localQuaternion.set(0, 0, 0, 1));
+          if (isCollide) {
+            voxel._risingState = 'colliding';
+          } else if (voxel._risingState === 'colliding') {
+            voxel._risingState = 'stopped';
+          }
+        }
+      });
+      this.voxelsY += 0.1;
+      if (this.voxelsY > this.highestY) {
+        this.rise2();
+      } else {
+        if (this.isAutoInit) this.update();
+      }
+    }
+    if (this.isRising2 && this.voxels2) {
+      // if (this.debugRender) console.log(this.getVoxel2(0, 0).position.y);
+      this.voxels2.children.forEach((voxel, i) => {
+        if (voxel._risingState === 'initial' || voxel._risingState === 'colliding') {
+          voxel.position.y += 0.1;
+          voxel.updateMatrixWorld();
+          // const isCollide = physicsManager.collideCapsule(0.5, 1, voxel.position, localQuaternion.set(0, 0, 0, 1), 1);
+          // const isCollide = physicsManager.collideBox(0.5, 0.05, 0.5, voxel.position, localQuaternion.set(0, 0, 0, 1), 1);
+          const isCollide = physicsManager.overlapBox(0.5, this.voxelHeightHalf, 0.5, voxel.position, localQuaternion.set(0, 0, 0, 1));
+          if (isCollide) {
+            voxel._risingState = 'colliding';
+          } else if (voxel._risingState === 'colliding') {
+            voxel._risingState = 'stopped';
+          }
+        }
+      });
+      this.voxelsY2 += 0.1;
+      if (this.voxelsY2 > this.highestY2) {
+        this.generateVoxelMap();
+        this.resolveInit(true);
+      } else {
+        if (this.isAutoInit) this.update();
+      }
+    }
+  }
+
+  generateVoxelMap() {
+    this.isRising = false;
+    this.isRising2 = false;
+
+    for (let z = -(this.height - 1) / 2; z < this.height / 2; z++) {
+      for (let x = -(this.width - 1) / 2; x < this.width / 2; x++) {
+        const currentVoxel = this.getVoxel(x, z);
+
+        const leftVoxel2 = this.getVoxel2(x - 1, z);
```

Comment:
This is a mess of `if` cases. Is this something that could be abstracted? Like...

```
leftVoxel2 = getVoxelDelta(-1, 0, 0);
rightVoxel2 = getVoxelDelta(1, 0, 0);
```